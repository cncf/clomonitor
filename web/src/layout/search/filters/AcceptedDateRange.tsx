import 'react-date-range/dist/styles.css'; // main style file
import 'react-date-range/dist/theme/default.css'; // theme css file

import classNames from 'classnames';
import moment from 'moment';
import { useEffect, useRef, useState } from 'react';
import { DateRange, Range, RangeKeyDict } from 'react-date-range';

import useOutsideClick from '../../../hooks/useOutsideClick';
import styles from './AcceptedDateRange.module.css';

interface Props {
  acceptedFrom?: string;
  acceptedTo?: string;
  onAcceptedDateRangeChange: (dates: AcceptedDate) => void;
}

interface AcceptedDate {
  accepted_from?: string;
  accepted_to?: string;
}

const INITIAL_DATE = '2016-01-01';

const AcceptedDateRange = (props: Props) => {
  const ref = useRef<HTMLDivElement>(null);
  const [showCalendar, setShowCalendar] = useState<boolean>(false);
  useOutsideClick([ref], showCalendar, () => setShowCalendar(false));

  const [ranges, setRanges] = useState<Range[]>([
    {
      startDate: moment(INITIAL_DATE).toDate(),
      endDate: moment().toDate(),
      key: 'selection',
    },
  ]);

  const handleChange = (data: RangeKeyDict) => {
    setRanges([data.selection]);
    props.onAcceptedDateRangeChange({
      accepted_from: moment(data.selection.startDate!).format('yyyy-MM-DD'),
      accepted_to: moment(data.selection.endDate!).format('yyyy-MM-DD'),
    });
  };

  useEffect(() => {
    setRanges([
      {
        startDate: moment(props.acceptedFrom || INITIAL_DATE).toDate(),
        endDate: moment(props.acceptedTo || moment()).toDate(),
        key: 'selection',
      },
    ]);
  }, [props.acceptedFrom, props.acceptedTo]); /* eslint-disable-line react-hooks/exhaustive-deps */

  return (
    <>
      <div className={`fw-bold text-uppercase text-primary ${styles.categoryTitle}`}>
        <small>Accepted</small>
      </div>

      <div className="mt-3">
        <button
          className={`btn btn-sm btn-outline-secondary rounded-0 w-100 ${styles.dateBtn}`}
          onClick={() => setShowCalendar(!showCalendar)}
          aria-label="Open calendar"
        >
          {moment(ranges[0].startDate!).format('MMM D, yyyy')} - {moment(ranges[0].endDate!).format('MMM D, yyyy')}
        </button>

        <div
          role="complementary"
          ref={ref}
          className={classNames(styles.dropdown, 'dropdown-menu tooltipDropdown rounded-0 text-wrap mt-2 p-0', {
            show: showCalendar,
          })}
        >
          <div className={`arrow ${styles.arrow}`} />

          <div className={`d-flex flex-row w-100 pt-3 text-uppercase text-muted ${styles.legends}`}>
            <div className={`w-50 ${styles.legendFrom}`}>From:</div>
            <div className={`w-50 ${styles.legendTo}`}>To:</div>
          </div>

          <DateRange
            className={styles.dateRange}
            editableDateInputs={true}
            onChange={handleChange}
            moveRangeOnFirstSelection={false}
            ranges={ranges}
            minDate={moment(INITIAL_DATE).toDate()}
            maxDate={moment().toDate()}
          />
        </div>
      </div>
    </>
  );
};

export default AcceptedDateRange;
