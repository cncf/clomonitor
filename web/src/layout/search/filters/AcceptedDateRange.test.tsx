import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import AcceptedDateRange from './AcceptedDateRange';

const mockOnAcceptedDateRangeChange = jest.fn();

const defaultProps = {
  onAcceptedDateRangeChange: mockOnAcceptedDateRangeChange,
};

describe('AcceptedDateRange', () => {
  let dateNowSpy: any;

  beforeEach(() => {
    dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => 1648154630000);
  });

  afterAll(() => {
    dateNowSpy.mockRestore();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<AcceptedDateRange {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders', () => {
      render(<AcceptedDateRange {...defaultProps} />);

      expect(screen.getAllByText('From:')).toHaveLength(2);
      expect(screen.getAllByText('To:')).toHaveLength(2);
      expect(screen.getByText('Jan 1, 2016')).toBeInTheDocument();
      expect(screen.getByText('Mar 24, 2022')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open calendar to choose date accepted_from' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open calendar to choose date accepted_to' })).toBeInTheDocument();

      const dropdowns = screen.getAllByRole('complementary');
      expect(dropdowns).toHaveLength(2);
      expect(dropdowns[0]).not.toHaveClass('show');
      expect(dropdowns[1]).not.toHaveClass('show');
    });

    it('displays calendar', async () => {
      render(<AcceptedDateRange {...defaultProps} />);

      const btn = screen.getByRole('button', { name: 'Open calendar to choose date accepted_from' });
      await userEvent.click(btn);

      expect(screen.getAllByRole('complementary')[0]).toHaveClass('show');
    });

    it('calls mockOnAcceptedDateRangeChange', async () => {
      render(<AcceptedDateRange {...defaultProps} />);

      const btn = screen.getByRole('button', { name: 'Open calendar to choose date accepted_from' });
      await userEvent.click(btn);

      expect(screen.getAllByRole('complementary')[0]).toHaveClass('show');

      const day20Btn = screen.getAllByText('20')[0].closest('button');
      await userEvent.click(day20Btn as HTMLButtonElement);

      expect(mockOnAcceptedDateRangeChange).toHaveBeenCalledTimes(1);
      expect(mockOnAcceptedDateRangeChange).toHaveBeenCalledWith({
        accepted_from: '2016-01-20',
      });
    });

    it('does not call to mockOnAcceptedDateRangeChange when range is the same than received props', async () => {
      render(<AcceptedDateRange {...defaultProps} acceptedFrom="2018-01-02" />);

      expect(mockOnAcceptedDateRangeChange).toHaveBeenCalledTimes(0);
    });
  });
});
