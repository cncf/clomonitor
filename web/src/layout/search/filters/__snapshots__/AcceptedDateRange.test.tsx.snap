// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcceptedDateRange creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="fw-bold text-uppercase text-primary categoryTitle"
  >
    <small>
      Accepted
    </small>
  </div>
  <div
    class="mt-3"
  >
    <button
      aria-label="Open calendar"
      class="btn btn-sm btn-outline-secondary rounded-0 w-100 dateBtn"
    >
      Jan 1, 2016 - Mar 24, 2022
    </button>
    <div
      class="dropdown dropdown-menu tooltipDropdown rounded-0 text-wrap mt-2 p-0"
      role="complementary"
    >
      <div
        class="arrow arrow"
      />
      <div
        class="d-flex flex-row w-100 pt-3 text-uppercase text-muted legends"
      >
        <div
          class="w-50 legendFrom"
        >
          From:
        </div>
        <div
          class="w-50 legendTo"
        >
          To:
        </div>
      </div>
      <div
        class="rdrCalendarWrapper rdrDateRangeWrapper dateRange"
      >
        <div
          class="rdrDateDisplayWrapper"
        >
          <div
            class="rdrDateDisplay"
            style="color: rgb(61, 145, 255);"
          >
            <span
              class="rdrDateInput rdrDateDisplayItem rdrDateDisplayItemActive"
            >
              <input
                placeholder="Early"
                value="Jan 1, 2016"
              />
            </span>
            <span
              class="rdrDateInput rdrDateDisplayItem"
            >
              <input
                placeholder="Continuous"
                value="Mar 24, 2022"
              />
            </span>
          </div>
        </div>
        <div
          class="rdrMonthAndYearWrapper"
        >
          <button
            class="rdrNextPrevButton rdrPprevButton"
            type="button"
          >
            <i />
          </button>
          <span
            class="rdrMonthAndYearPickers"
          >
            <span
              class="rdrMonthPicker"
            >
              <select>
                <option
                  value="0"
                >
                  January
                </option>
                <option
                  value="1"
                >
                  February
                </option>
                <option
                  value="2"
                >
                  March
                </option>
                <option
                  value="3"
                >
                  April
                </option>
                <option
                  value="4"
                >
                  May
                </option>
                <option
                  value="5"
                >
                  June
                </option>
                <option
                  value="6"
                >
                  July
                </option>
                <option
                  value="7"
                >
                  August
                </option>
                <option
                  value="8"
                >
                  September
                </option>
                <option
                  value="9"
                >
                  October
                </option>
                <option
                  value="10"
                >
                  November
                </option>
                <option
                  value="11"
                >
                  December
                </option>
              </select>
            </span>
            <span />
            <span
              class="rdrYearPicker"
            >
              <select>
                <option
                  value="2022"
                >
                  2022
                </option>
                <option
                  value="2021"
                >
                  2021
                </option>
                <option
                  value="2020"
                >
                  2020
                </option>
                <option
                  value="2019"
                >
                  2019
                </option>
                <option
                  value="2018"
                >
                  2018
                </option>
                <option
                  value="2017"
                >
                  2017
                </option>
                <option
                  value="2016"
                >
                  2016
                </option>
              </select>
            </span>
          </span>
          <button
            class="rdrNextPrevButton rdrNextButton"
            type="button"
          >
            <i />
          </button>
        </div>
        <div
          class="rdrMonths rdrMonthsVertical"
        >
          <div
            class="rdrMonth"
          >
            <div
              class="rdrWeekDays"
            >
              <span
                class="rdrWeekDay"
              >
                Sun
              </span>
              <span
                class="rdrWeekDay"
              >
                Mon
              </span>
              <span
                class="rdrWeekDay"
              >
                Tue
              </span>
              <span
                class="rdrWeekDay"
              >
                Wed
              </span>
              <span
                class="rdrWeekDay"
              >
                Thu
              </span>
              <span
                class="rdrWeekDay"
              >
                Fri
              </span>
              <span
                class="rdrWeekDay"
              >
                Sat
              </span>
            </div>
            <div
              class="rdrDays"
            >
              <button
                class="rdrDay rdrDayPassive rdrDayDisabled rdrDayWeekend rdrDayStartOfWeek"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    27
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive rdrDayDisabled"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    28
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive rdrDayDisabled"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    29
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive rdrDayDisabled"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    30
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive rdrDayDisabled"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    31
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayStartOfMonth"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrStartEdge"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    1
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayEndOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    2
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayStartOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    3
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    4
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    5
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    6
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    7
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    8
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayEndOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    9
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayStartOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    10
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    11
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    12
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    13
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    14
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    15
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayEndOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    16
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayStartOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    17
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    18
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    19
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    20
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    21
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    22
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayEndOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    23
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayStartOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    24
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    25
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    26
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    27
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    28
                  </span>
                </span>
              </button>
              <button
                class="rdrDay"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    29
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayEndOfWeek"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    30
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayWeekend rdrDayStartOfWeek rdrDayEndOfMonth"
                style="color: rgb(61, 145, 255);"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    31
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    1
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    2
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    3
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    4
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    5
                  </span>
                </span>
              </button>
              <button
                class="rdrDay rdrDayPassive rdrDayWeekend rdrDayEndOfWeek"
                style="color: rgb(61, 145, 255);"
                tabindex="-1"
                type="button"
              >
                <span
                  class="rdrInRange"
                  style="color: rgb(61, 145, 255);"
                />
                <span
                  class="rdrDayNumber"
                >
                  <span>
                    6
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
